File: ui/pages/login_page.py
from playwright.sync_api import Locator, Page

from pages.base_page import BasePage


class LoginPage(BasePage):
    """Actions/selectors for the SauceDemo login screen."""

    PATH = "/"

    def __init__(self, page: Page, base_url: str) -> None:
        super().__init__(page, base_url)
        self._username: Locator = page.locator("#user-name")
        self._password: Locator = page.locator("#password")
        self._submit: Locator = page.locator("#login-button")

    def open(self) -> None:
        self.goto(self.PATH)

    def login(self, username: str, password: str) -> None:
        # Single login entry-point used by positive and negative auth tests.
        self._username.fill(username)
        self._password.fill(password)
        self._submit.click()

    @property
    def error_message(self) -> Locator:
        return self.page.locator("[data-test='error']")

File: java/playwright/src/test/java/com/tnd/playwright/pages/LoginPage.java
package com.tnd.playwright.pages;

import com.microsoft.playwright.Page;

/**
 * Page object for SauceDemo login screen (Playwright Java).
 *
 * <p>Purpose: encapsulate auth-related UI actions used by all flow tests.
 */
public class LoginPage extends BasePage {
    /**
     * Creates login page object bound to current browser page and base URL.
     */
    public LoginPage(Page page, String baseUrl) {
        super(page, baseUrl);
    }

    /**
     * Opens the login entry route.
     */
    public void open() {
        goTo("/");
    }

    /**
     * Performs credential submission for positive/negative auth scenarios.
     */
    public void login(String username, String password) {
        // Shared login action used by all core flow tests.
        page.locator("#user-name").fill(username);
        page.locator("#password").fill(password);
        page.locator("#login-button").click();
    }
}
